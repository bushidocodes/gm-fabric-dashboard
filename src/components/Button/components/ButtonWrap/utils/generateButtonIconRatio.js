/**
 * Helper function that maps button sizes to numbers representing icon ratios
 *
 * @param {string} size
 * @returns number
 */
function mapSizeToButtonIconRatio(size) {
  switch (size) {
    case "xs":
      return 0.5;
    case "sm":
      return 0.75;
    case "lg":
      return 1.25;
    case "xl":
      return 1.5;
    case "normal":
    default:
      return 1;
  }
}

/**
 * Utility function that generates a CSS mixin containing button icon rulesets
 *
 * @param {string} size
 * @returns string
 */
function generateButtonIconRatio(size) {
  const viewboxSize = 24; // default viewbox size 24px
  const ratio = mapSizeToButtonIconRatio(size); // default ratio 1
  const viewboxDimensions = viewboxSize * ratio;
  return `> svg { width: ${/* sc-value */ viewboxDimensions}px; 
  height: ${/* sc-value */ viewboxDimensions}px; }`;
}

export default generateButtonIconRatio;
