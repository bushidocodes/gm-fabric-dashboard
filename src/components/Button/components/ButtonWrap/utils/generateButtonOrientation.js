import { spacingScale } from "style/styleFunctions";

/**
 * Utility function that generates a CSS mixin containing button orientation rulesets
 *
 * @param {string} [alignment="horizontal"]
 * @param {number} [spacingBase=0.5]
 * @returns string
 */
function generateButtonOrientation(
  alignment = "horizontal",
  spacingBase = 0.5
) {
  if (alignment === "horizontal") {
    return `
      flex-direction: row;
      > :not(:first-child) {
        margin-left: ${/* sc-value */ spacingScale(spacingBase)};
        flex-basis: auto;
      }
    `;
  } else if (alignment === "vertical") {
    return `
      flex-direction: column;
      padding-top: ${/* sc-value */ spacingScale(spacingBase)};
      > :not(:first-child) {
        margin-top: ${/* sc-value */ spacingBase / 2}px;
      }
    `;
  }
}

export default generateButtonOrientation;
