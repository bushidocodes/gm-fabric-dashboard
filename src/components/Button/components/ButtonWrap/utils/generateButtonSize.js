import { spacingScale } from "style/styleFunctions";

/**
 * Utility function that transforms an object containing different attributes
 * into a block of CSS attributes dealing with button size
 *
 * @param {any} {
 *   fontSizeMultiplier = 1,
 *   horizontalPaddingMultiplier = 1,
 *   verticalPaddingMultiplier = 1,
 *   textIsUppercase = false
 * }
 * @returns string
 */
function _buttonSizeStyleBlockGenerator({
  fontSizeMultiplier = 1,
  horizontalPaddingMultiplier = 1,
  verticalPaddingMultiplier = 1,
  textIsUppercase = false
}) {
  const fontSizeBase = 14;
  const borderRadiusBase = 3;
  const contentSpacingBase = parseInt(spacingScale(0.5), 10);
  return `
    font-size: ${Math.ceil(fontSizeBase * fontSizeMultiplier)}px;
    border-radius: ${borderRadiusBase}px;
    padding: ${contentSpacingBase *
      horizontalPaddingMultiplier}px ${contentSpacingBase *
    verticalPaddingMultiplier}px;
    ${textIsUppercase ? "text-transform: uppercase;" : ""}
  `;
}

/**
 * Utility function that renders a block of CSS attributes for a
 * particular string (xs, sm, normal, lg, xl)
 *
 * @param {string} size
 * @returns
 */
function generateButtonSize(size) {
  switch (size) {
    case "xs": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 0.5,
        horizontalPaddingMultiplier: 0.25,
        verticalPaddingMultiplier: 1,
        textIsUppercase: true
      });
    }
    case "sm": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 0.75,
        horizontalPaddingMultiplier: 0.25,
        verticalPaddingMultiplier: 1,
        textIsUppercase: true
      });
    }
    case "lg": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 1.25,
        horizontalPaddingMultiplier: 0.5,
        verticalPaddingMultiplier: 1,
        textIsUppercase: false
      });
    }
    case "xl": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 1.5,
        horizontalPaddingMultiplier: 0.5,
        verticalPaddingMultiplier: 1,
        textIsUppercase: false
      });
    }
    default:
    case "normal": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 1,
        horizontalPaddingMultiplier: 0.35,
        verticalPaddingMultiplier: 1,
        textIsUppercase: false
      });
    }
  }
}

export default generateButtonSize;
