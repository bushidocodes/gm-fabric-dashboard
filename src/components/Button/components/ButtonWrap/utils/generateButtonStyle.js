import { contrastColor } from "style/styleFunctions";
import { COLOR_BRAND_PRIMARY, COLOR_WHITE } from "style/styleVariables";
import { darkenColor, lightenColor } from "style/styleFunctions";

/**
 * Utility function that transforms an object containing different attributes
 * into a block of CSS attributes dealing with buton styles
 *
 * @param {any} {
 *   buttonBackgroundColorBase = COLOR_WHITE,
 *   buttonBorderColorBase = COLOR_WHITE,
 *   buttonLabelColorBase = contrastColor(COLOR_WHITE, 100),
 *   buttonActiveStatus = false,
 *   buttonOutlineStyle = "raisedOutline",
 *   buttonOutlineStyleDepth = 0.06, // 0-100% // default: 6%,
 *   buttonReactionStyle = "darken", // darken, lighten // default: darken
 *   buttonReactionDegree = 0.06, // 0-100% // default: 6%
 *   buttonHoverReactionDegree = 0.06 * 0.25,
 *   buttonActiveReactionDegree = 0.06 * 2,
 *   buttonDownReactionDegree = 0.06,
 *   buttonActiveMixBaseColor = COLOR_BRAND_PRIMARY,
 *   buttonActiveMixBaseLabelColor = contrastColor(COLOR_BRAND_PRIMARY, 100),
 *   buttonActiveMixDegree = 1 //default 100%
 * }
 * @returns string
 */
function generateButtonStyle({
  buttonBackgroundColorBase = COLOR_WHITE,
  buttonBorderColorBase = COLOR_WHITE,
  buttonLabelColorBase = contrastColor(COLOR_WHITE, 100),
  buttonActiveStatus = false,
  buttonOutlineStyle = "raisedOutline",
  buttonOutlineStyleDepth = 0.06, // 0-100% // default: 6%,
  buttonReactionStyle = "darken", // darken, lighten // default: darken
  buttonReactionDegree = 0.06, // 0-100% // default: 6%
  buttonHoverReactionDegree = 0.06 * 0.25,
  buttonActiveReactionDegree = 0.06 * 2,
  buttonDownReactionDegree = 0.06,
  buttonActiveMixBaseColor = COLOR_BRAND_PRIMARY,
  buttonActiveMixBaseLabelColor = contrastColor(COLOR_BRAND_PRIMARY, 100),
  buttonActiveMixDegree = 1 //default 100%
}) {
  // default style Types (static, hover, down, active, activeDown)
  // this object feeds all the css at the bottom of the function
  // in this object we unite defaults with inputs
  let styleTypes = {
    staticStyles: {
      backgroundColor: buttonBackgroundColorBase,
      color: buttonLabelColorBase,
      borderColor: buttonBorderColorBase
    },
    hoverStyles: {},
    downStyles: {},
    activeStyles: {},
    activeDownStyles: {}
  };

  // Create style for each property, in each state, under each processed condition
  if (buttonReactionStyle === "darken") {
    // Hover Styles
    styleTypes.hoverStyles["backgroundColor"] = darkenColor(
      buttonBackgroundColorBase,
      buttonHoverReactionDegree
    );
    styleTypes.hoverStyles["color"] = darkenColor(
      buttonLabelColorBase,
      buttonHoverReactionDegree
    );
    styleTypes.hoverStyles["borderColor"] = darkenColor(
      buttonBorderColorBase,
      buttonHoverReactionDegree
    );
    // Down Styles
    styleTypes.downStyles["backgroundColor"] = darkenColor(
      buttonBackgroundColorBase,
      buttonHoverReactionDegree + buttonDownReactionDegree
    );
    styleTypes.downStyles["color"] = darkenColor(
      buttonLabelColorBase,
      buttonHoverReactionDegree + buttonDownReactionDegree
    );
    styleTypes.downStyles["borderColor"] = darkenColor(
      buttonBorderColorBase,
      buttonHoverReactionDegree + buttonDownReactionDegree
    );
    // Active Styles
    styleTypes.activeStyles["backgroundColor"] = darkenColor(
      buttonBackgroundColorBase,
      buttonActiveReactionDegree
    ).mix(buttonActiveMixBaseColor, buttonActiveMixDegree);
    styleTypes.activeStyles["color"] = darkenColor(
      buttonLabelColorBase,
      buttonActiveReactionDegree
    ).mix(buttonActiveMixBaseLabelColor, buttonActiveMixDegree);
    styleTypes.activeStyles["borderColor"] = darkenColor(
      buttonBorderColorBase.mix(
        buttonActiveMixBaseColor,
        buttonActiveMixDegree
      ),
      buttonActiveReactionDegree + buttonDownReactionDegree
    );
    // Active Down Styles
    styleTypes.activeDownStyles["backgroundColor"] = darkenColor(
      buttonBackgroundColorBase,
      buttonActiveReactionDegree + buttonDownReactionDegree
    ).mix(buttonActiveMixBaseColor, buttonActiveMixDegree);
    styleTypes.activeDownStyles["color"] = darkenColor(
      buttonLabelColorBase,
      buttonActiveReactionDegree + buttonDownReactionDegree
    ).mix(buttonActiveMixBaseLabelColor, buttonActiveMixDegree);
    styleTypes.activeDownStyles["borderColor"] = darkenColor(
      buttonBorderColorBase,
      buttonActiveReactionDegree + buttonDownReactionDegree
    ).mix(buttonActiveMixBaseColor, buttonActiveMixDegree);
  } else {
    // Assume button reaction style == lighten
    // Hover Styles
    styleTypes.hoverStyles["backgroundColor"] = lightenColor(
      buttonBackgroundColorBase,
      buttonHoverReactionDegree
    );
    styleTypes.hoverStyles["color"] = lightenColor(
      buttonLabelColorBase,
      buttonHoverReactionDegree
    );
    styleTypes.hoverStyles["borderColor"] = lightenColor(
      buttonBorderColorBase,
      buttonHoverReactionDegree
    );
    // Down Styles
    styleTypes.downStyles["backgroundColor"] = lightenColor(
      buttonBackgroundColorBase,
      buttonHoverReactionDegree + buttonDownReactionDegree
    );
    styleTypes.downStyles["color"] = lightenColor(
      buttonLabelColorBase,
      buttonHoverReactionDegree + buttonDownReactionDegree
    );
    styleTypes.downStyles["borderColor"] = lightenColor(
      buttonBorderColorBase,
      buttonHoverReactionDegree + buttonDownReactionDegree
    );
    // Active Styles
    styleTypes.activeStyles["backgroundColor"] = lightenColor(
      buttonBackgroundColorBase,
      buttonActiveReactionDegree
    ).mix(buttonActiveMixBaseColor, buttonActiveMixDegree);
    styleTypes.activeStyles["color"] = lightenColor(
      buttonLabelColorBase,
      buttonActiveReactionDegree
    ).mix(buttonActiveMixBaseLabelColor, buttonActiveMixDegree);
    styleTypes.activeStyles["borderColor"] = lightenColor(
      buttonBorderColorBase,
      buttonActiveReactionDegree
    ).mix(buttonActiveMixBaseColor, buttonActiveMixDegree);
    // Active Down Styles
    styleTypes.activeDownStyles["backgroundColor"] = lightenColor(
      buttonBackgroundColorBase,
      buttonActiveReactionDegree + buttonDownReactionDegree
    ).mix(buttonActiveMixBaseColor, buttonActiveMixDegree);
    styleTypes.activeDownStyles["color"] = lightenColor(
      buttonLabelColorBase,
      buttonActiveReactionDegree + buttonDownReactionDegree
    ).mix(buttonActiveMixBaseLabelColor, buttonActiveMixDegree);
    styleTypes.activeDownStyles["borderColor"] = lightenColor(
      buttonBorderColorBase,
      buttonActiveReactionDegree + buttonDownReactionDegree
    ).mix(buttonActiveMixBaseColor, buttonActiveMixDegree);
  }

  // Generation of outline styles based on styleOptions and styleTypes objects
  const buttonOutlineStyles = {
    raisedOutline: {
      staticStyles: {
        borderColor: darkenColor(
          styleTypes.staticStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderTopColor: lightenColor(
          styleTypes.staticStyles.borderColor,
          buttonOutlineStyleDepth * 0.5
        ),
        borderBottomColor: darkenColor(
          styleTypes.staticStyles.borderColor,
          buttonOutlineStyleDepth * 3
        )
      },
      hoverStyles: {
        borderColor: darkenColor(
          styleTypes.hoverStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderTopColor: lightenColor(
          styleTypes.hoverStyles.borderColor,
          buttonOutlineStyleDepth * 0.5
        ),
        borderBottomColor: darkenColor(
          styleTypes.hoverStyles.borderColor,
          buttonOutlineStyleDepth * 3
        )
      },
      downStyles: {
        borderColor: darkenColor(
          styleTypes.downStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderTopColor: lightenColor(
          styleTypes.downStyles.borderColor,
          buttonOutlineStyleDepth * 0.5
        ),
        borderBottomColor: darkenColor(
          styleTypes.downStyles.borderColor,
          buttonOutlineStyleDepth * 3
        )
      },
      activeStyles: {
        borderColor: darkenColor(
          styleTypes.activeStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderTopColor: lightenColor(
          styleTypes.activeStyles.borderColor,
          buttonOutlineStyleDepth * 0.5
        ),
        borderBottomColor: darkenColor(
          styleTypes.activeStyles.borderColor,
          buttonOutlineStyleDepth * 3
        )
      },
      activeDownStyles: {
        borderColor: darkenColor(
          styleTypes.activeDownStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderTopColor: lightenColor(
          styleTypes.activeDownStyles.borderColor,
          buttonOutlineStyleDepth * 0.5
        ),
        borderBottomColor: darkenColor(
          styleTypes.activeDownStyles.borderColor,
          buttonOutlineStyleDepth * 3
        )
      }
    },
    shadow: {
      staticStyles: {
        borderBottomColor: darkenColor(
          styleTypes.staticStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      hoverStyles: {
        borderBottomColor: darkenColor(
          styleTypes.hoverStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      downStyles: {
        borderBottomColor: darkenColor(
          styleTypes.downStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      activeStyles: {
        borderBottomColor: darkenColor(
          styleTypes.activeStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      activeDownStyles: {
        borderBottomColor: darkenColor(
          styleTypes.activeDownStyles.borderColor,
          buttonOutlineStyleDepth
        )
      }
    },
    raised: {
      staticStyles: {
        borderTopColor: lightenColor(
          styleTypes.staticStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderBottomColor: darkenColor(
          styleTypes.staticStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      hoverStyles: {
        borderTopColor: lightenColor(
          styleTypes.hoverStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderBottomColor: darkenColor(
          styleTypes.hoverStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      downStyles: {
        borderTopColor: lightenColor(
          styleTypes.downStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderBottomColor: darkenColor(
          styleTypes.downStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      activeStyles: {
        borderTopColor: lightenColor(
          styleTypes.activeStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderBottomColor: darkenColor(
          styleTypes.activeStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      activeDownStyles: {
        borderTopColor: lightenColor(
          styleTypes.activeDownStyles.borderColor,
          buttonOutlineStyleDepth
        ),
        borderBottomColor: darkenColor(
          styleTypes.activeDownStyles.borderColor,
          buttonOutlineStyleDepth
        )
      }
    },
    outline: {
      staticStyles: {
        borderColor: darkenColor(
          styleTypes.staticStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      hoverStyles: {
        borderColor: darkenColor(
          styleTypes.hoverStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      downStyles: {
        borderColor: darkenColor(
          styleTypes.downStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      activeStyles: {
        borderColor: darkenColor(
          styleTypes.activeStyles.borderColor,
          buttonOutlineStyleDepth
        )
      },
      activeDownStyles: {
        borderColor: darkenColor(
          styleTypes.activeDownStyles.borderColor,
          buttonOutlineStyleDepth
        )
      }
    },
    none: {
      staticStyles: {
        borderColor: styleTypes.staticStyles.backgroundColor
      },
      activeStyles: {
        borderColor: styleTypes.activeStyles.backgroundColor
      }
    }
  };

  // Distribute generated outline styles based on the input
  // by overwriting default generated outline styles
  const outlineOptionsTypes = buttonOutlineStyles[buttonOutlineStyle];
  const outlineOptionTypesKeys = Object.keys(outlineOptionsTypes);
  // Loop through generated buttonOutlineStyles that correspond to the input outline Style (raised, raisedOutline, none, shadow, etc.)
  for (let x = 0, len = outlineOptionTypesKeys.length; x < len; x++) {
    // Then look into each Type (static, active, activeDown) within
    // styleTypes object (which is the output that feeds the css styles at the bottom)
    // and either supplement or overwrite existing properties
    // based on the information that is within generated outline styles
    let outlineOptionKeys = Object.keys(
      outlineOptionsTypes[outlineOptionTypesKeys[x]]
    );
    for (let y = 0, len = outlineOptionKeys.length; y < len; y++) {
      styleTypes[outlineOptionTypesKeys[x]][outlineOptionKeys[y]] =
        outlineOptionsTypes[outlineOptionTypesKeys[x]][outlineOptionKeys[y]];
    }
  }

  const baseStyles = buttonActiveStatus
    ? styleTypes.activeStyles
    : styleTypes.staticStyles;
  const downStyles = buttonActiveStatus
    ? styleTypes.activeDownStyles
    : styleTypes.downStyles;
  const hoverStyles = buttonActiveStatus
    ? styleTypes.activeStyles
    : styleTypes.hoverStyles;

  return `
  background-color: ${baseStyles["backgroundColor"].string()};
  border-color: ${baseStyles["borderColor"].string()};
  ${
    baseStyles["borderTopColor"]
      ? `border-top-color: ${baseStyles["borderTopColor"].string()};`
      : ""
  }
  ${
    baseStyles["borderBottomColor"]
      ? `border-bottom-color: ${baseStyles["borderBottomColor"].string()};`
      : ""
  }
  color: ${baseStyles["color"].string()};
  &:active {
    ${!buttonActiveStatus ? "transition-duration: 0s;" : ""}
    background-color: ${downStyles["backgroundColor"].string()};
    border-color: ${downStyles["borderColor"].string()};
    ${
      downStyles["borderTopColor"]
        ? `border-top-color: ${downStyles["borderTopColor"].string()};`
        : ""
    }
    ${
      downStyles["borderBottomColor"]
        ? `border-bottom-color: ${downStyles["borderBottomColor"].string()};`
        : ""
    }
    color: ${downStyles["color"].string()};
  }

  &:hover {
    background-color: ${hoverStyles["backgroundColor"].string()};
    border-color: ${hoverStyles["borderColor"].string()};
    ${
      hoverStyles["borderTopColor"]
        ? `border-top-color: ${hoverStyles["borderTopColor"].string()};`
        : ""
    }
    ${
      hoverStyles["borderBottomColor"]
        ? `border-bottom-color: ${hoverStyles["borderBottomColor"].string()};`
        : ""
    }
    color: ${hoverStyles["color"].string()};
  } 

  &:focus {
    outline: none;
    box-shadow: 0 0 0 2px #0aab2a;
    z-index: 1;
  }

  &[disabled],
  &[disabled]:hover,
  &[disabled]:focus,
  &[disabled]:active {
    cursor: default;
    opacity: 0.8;
    background-color: ${darkenColor(
      buttonBackgroundColorBase.desaturate(1),
      buttonReactionDegree
    ).string()};
    border-color: ${darkenColor(
      buttonBorderColorBase.desaturate(1),
      buttonReactionDegree
    ).string()};
    color: ${darkenColor(
      buttonLabelColorBase.desaturate(1),
      buttonReactionDegree
    ).string()};
  }
`;
}

export default generateButtonStyle;
