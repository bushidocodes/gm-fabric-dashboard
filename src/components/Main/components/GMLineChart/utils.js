// importing a single function instead of the whole _
import isEmpty from "lodash/isEmpty";

/**
 * Converts timeseries with time stamps to a string of values without timestamps
 *
 * @param {array} array a timeseries array with time stamps
 * @returns {array}     an array of values
 */
export function numericalTimeSeriesFunc(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array[0].map(cell => cell[1]);
  } else {
    return 0;
  }
}

/**
 * Produces an average of array of values
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function average(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array.reduce((a, b) => a + b) / array.length;
  } else {
    return 0;
  }
}

/**
 * Produces a median of array of values
 * (if no middle value, average of 2 most central values is returned)
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function median(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    const len = array.length;
    let median = 0;
    array.sort();

    if (len % 2 === 0) {
      // if even, average of two middle numbers
      median = (array[len / 2 - 1] + array[len / 2]) / 2;
    } else {
      // if odd, middle number only
      median = array[(len - 1) / 2];
    }
    return median;
  } else {
    return 0;
  }
}

/**
 * Produces a mode(s) of array of values
 * if more than one mode, returns all applicable modes as well as their frequency
 *
 * @param {array} array an array of values
 * @returns {array}     an array that contains mode(s) and frequency value
 */
export function modes(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    let modes = [];
    let count = [];
    let i;
    let number;
    let maxIndex = 0;

    for (let i = 0; i < array.length; i++) {
      number = array[i];
      count[number] = (count[number] || 0) + 1;
      if (count[number] > maxIndex) {
        maxIndex = count[number];
      }
    }

    for (i in count)
      if (count.hasOwnProperty(i)) {
        if (count[i] === maxIndex) {
          modes.push(Number(i));
        }
      }

    modes.push({ frequency: maxIndex });

    if (maxIndex > 1) {
      return modes;
    } else {
      modes = "no value occurs more than once";
      return modes;
    }
  } else {
    return 0;
  }
}

/**
 * Produces a string that contains descriptive statistics of the graph
 *
 * @param {array} timeSeries an array of values
 * @param {string} title chart title
 * @returns {string}     returns a string with summary chart statistics
 */
export function screenReaderGraphDescription(timeSeries, title, intl) {
  const numericalTimeSeries = numericalTimeSeriesFunc(timeSeries);

  // We don't have negative values in our charts, thus average of 0 means that all values are 0, aka no meaningful data
  if (average(numericalTimeSeries) > 0) {
    let screenReaderGraphData = {
      median: median(numericalTimeSeries).toFixed(2),
      average: average(numericalTimeSeries).toFixed(2),
      mode: modes(numericalTimeSeries),
      max: numericalTimeSeries.sort()[numericalTimeSeries.length - 1],
      min: numericalTimeSeries.sort()[0],
      dataPoints: numericalTimeSeries.length,
      timeSeries: numericalTimeSeries,
      title
    };

    if (
      typeof screenReaderGraphData.mode !== "string" &&
      screenReaderGraphData.mode.length > 1
    ) {
      screenReaderGraphData = Object.assign({}, screenReaderGraphData, {
        mode: `Mode has frequency of ${
          screenReaderGraphData.mode.pop().frequency
        } and values of ${screenReaderGraphData.mode}`
      });
    }

    return intl.formatMessage(
      {
        id: "GMLineChart.screenReaderGraphDescription.withData",
        defaultMessage:
          "A tabular representation of the {title} chart data: median {median} average {average} mode {mode} maximum {max} minimum {min} number of observations {dataPoints} complete data time series follows {timeSeries}",
        description: "Screen reader description for GMLineChart"
      },
      screenReaderGraphData
    );
  } else {
    return intl.formatMessage({
      id: "GMLineChart.screenReaderGraphDescription.noData",
      defaultMessage: `The average for currently displayed data is equal to 0.`,
      description: "Screen reader description for GMLineChart"
    });
  }
}
