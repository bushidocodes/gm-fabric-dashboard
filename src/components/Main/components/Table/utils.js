import React from "react";
import TableColHeader from "components/Main/components/TableColHeader";
import TableColLatencyHeader from "components/Main/components/TableColLatencyHeader";

/** function getTableHeaders
 * takes Table type and returns appropriate headers with styling
 * for Function/Routes Table or GMServiceTable (Instances Table)
 * @param {object} items
 * @returns {React.element}
 */
export const getTableHeaders = type => {
  let headerNames = [];
  if (type === "Instance") headerNames = [type, "Uptime"];
  if (type === "Route" || type === "Function")
    headerNames = [type, "Requests/sec", "Requests", "Error %", "Latency"];
  let styleProps = {};

  // apply greater flex-basis for first column (20%)
  // requests/sec is a sparkline header and set to align left
  // set text-align to right for additional text columns
  return headerNames.map((header, headerIdx) => {
    styleProps = {};
    if (headerIdx === 0) styleProps = { flex: "1 1 20%" };
    if (headerIdx !== 0 && header !== "Requests/sec")
      styleProps = { textAlign: "right" };

    if (header === "Latency") {
      return <TableColLatencyHeader key={header} />;
    } else {
      return (
        <TableColHeader style={styleProps} key={header}>
          {header}
        </TableColHeader>
      );
    }
  });
};

// simple helper that evaluates and returns route or func props
export const getItem = (route, func) => {
  if (route !== undefined) return route;
  if (func !== undefined) return func;
};
