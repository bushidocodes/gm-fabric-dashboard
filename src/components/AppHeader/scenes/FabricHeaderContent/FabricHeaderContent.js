import React from "react";
import { PropTypes } from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { microserviceStatuses } from "utils/constants";
import { getStatusCount } from "utils/selectors";

import Tab from "components/AppHeader/components/Tab";
import TabNav from "components/AppHeader/components/TabNav";

FabricAppHeaderContent.propTypes = {
  statusCount: PropTypes.shape({
    Down: PropTypes.number,
    Stable: PropTypes.number,
    Warning: PropTypes.number,
    total: PropTypes.number.isRequired
  })
};

function FabricAppHeaderContent({ statusCount }) {
  return (
    <TabNav>
      <Tab
        title="All Services"
        href="/"
        icon="Summary"
        lines={[
          {
            name: "Services",
            value: statusCount.total
          }
        ]}
      />
      {microserviceStatuses.map(status => {
        return (
          <Tab
            title={status}
            href={`/${status}`}
            icon={status}
            lines={[
              {
                name: "Services",
                value: statusCount[status]
              }
            ]}
            key={status}
          />
        );
      })}
    </TabNav>
  );
}

function mapStateToProps(state) {
  return {
    statusCount: getStatusCount(state)
  };
}

export default withRouter(connect(mapStateToProps)(FabricAppHeaderContent));
