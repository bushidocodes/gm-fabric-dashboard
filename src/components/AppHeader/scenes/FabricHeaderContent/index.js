import FabricHeaderContent from "./FabricHeaderContent";
export default FabricHeaderContent;
