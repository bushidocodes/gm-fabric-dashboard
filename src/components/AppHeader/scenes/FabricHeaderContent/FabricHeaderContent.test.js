import React from "react";
import { MemoryRouter, Route } from "react-router";
import configureMockStore from "redux-mock-store";

import { renderWithIntl } from "utils/i18nTesting";
import state from "json/mockReduxState";
import FabricHeaderContent from "./FabricHeaderContent";

// Router is necessary because of the <Tab />'s which need a router context
const RouterWrap = (
  <MemoryRouter>
    <Route
      render={props => (
        <FabricHeaderContent {...props} store={configureMockStore()(state)} />
      )}
    />
  </MemoryRouter>
);

describe("FabricHeaderContent", () => {
  test("matches snapshot", () => {
    const tree = renderWithIntl(RouterWrap);
    expect(tree).toMatchSnapshot();
  });
});
