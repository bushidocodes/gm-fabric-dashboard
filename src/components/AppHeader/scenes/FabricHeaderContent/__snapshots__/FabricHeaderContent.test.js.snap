// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FabricHeaderContent matches snapshot 1`] = `null`;
